### ISSUE TITLE
FR: A simple 4-digits pin 

### ISSUE THREAD

**Issue by ironhak** [koreader/koreader#10912]
**Describe the solution you'd like**
A simple 4-digits passcode in order to unlock the device, just like the original Kindle and Kobo software. 

**Additional context**
I think this would be very important for many reasons:
- I don't want anyone to be able to open my device and see what I'm reading. 
- KOReader is way more advanced than stock software (Kobo & Kindle) but lacks of this very basic feature. 
- I have a passcode on all my device (as everyone I think), I just don't like to have a device easily accessible by everyone. I want my device to be accessible only by me. 

This is the only thing preventing me to use KOReader on my Kobo, for sure other people feel the same.

**Comment by poire-z** [comment]
We don't have 4-digit pin code, but we have an option to require a specific gesture to get out of screensaver.
It's a bit hidden, `Gear > Screen > Screensaver > Settings > Keep the screensaver on screen after wakeup > Until 'Exit screensaver' gesture` (phew....)
And in `Tap & gestures > Gesture manager`, you can set the action "Exit screensaver" to the gesture of your choice (including a twisted one with Custom multiswipes).

**Comment by ironhak** [comment]
> We don't have 4-digit pin code, but we have an option to require a specific gesture to get out of screensaver.
> It's a bit hidden, `Gear > Screen > Screensaver > Settings > Keep the screensaver on screen after wakeup > Until 'Exit screensaver' gesture` (phew....)
> And in `Tap & gestures > Gesture manager`, you can set the action "Exit screensaver" to the gesture of your choice (including a twisted one with Custom multiswipes).

I know, unfortunately it does not work on my device (don’t know why). Also, why not just implement a 4 digit code? I mean it was surely more complex implementing the multi-swipe feature than it would be a simple 4-digits passcode 😅

**Comment by NiLuJe** [comment]
> I mean it was surely more complex implementing the multi-swipe feature than it would be a simple 4-digits passcode 😅

And you'd be wrong ;). We've already discussed this in the past (it's an old issue that was probably linked to in the gesture PR), but that approach has a very large amount of tiny edge-cases that are extremely annoying to deal with.

----

I'm mildly interested in why you can't get the gesture to work, though. Verbose debug logs might shed some light on this.

**Comment by Frenzie** [comment]
There's also the factor that multiswipes were a thing that interested me, so I wrote them. That's mostly how development works around here. :-)

**Comment by ironhak** [comment]
So it will never be implemented?

**Comment by Frenzie** [comment]
If you mean never by me then probably, but I don't think there are any big objections to it.

**Comment by poire-z** [comment]
> So it will never be implemented?

Probably not: it would be quite more complicated than the gesture unlock we have.
We have that gesture unlock feature for people like you who needs privacy, so we'd rather help you having it working than spending time adding kludge to display a working textinput + a keyboard over a lock reader.

**Comment by mergen3107** [comment]
> I know, unfortunately it does not work on my device (don’t know why).

Recently I asked how to setup and use custom multiswipe for unlocking KOReader in #11224. 

Summary:
1) Gesture unlock should setup in Gesture manager from both File Manager and Reader.
2) Custom multiswipes only support one type of gestures in a single multiswipe. So, only straight ones (up/down/left/right) or only diagonal ones.
3) Try setting up a gesture you you know will work (corner tap, or double finger swipe).
4) Then make a custom multiswipe, then practice on a simple action like toggle frontlight to get used to it. Then switch it to Exit screensaver action (in FM and Reader).

**Comment by ironhak** [comment]
Please guys... It's 2024. Costs nothing to insert a simple 4 digits pin. Please

**Comment by Commodore64user** [comment]
> Please guys... It's 2024. Costs nothing to insert a simple 4 digits pin. Please

it's 2024, everything has a costs somehow, especially in our [mostly] capitalist world. Time, effort, opportunity cost.

**Comment by Menghini** [comment]
I agree with this. This would also help accidental unlocks. However, the `Exit screensaver gesture` will work for me in it's place.

**Comment by Commodore64user** [comment]
As a matter of courtesy (and interest), could we link to the previous discussion or mention the difficulties with implementing said FR?

**Comment by jonnyl2** [comment]
Here's another related thread: #5970

**Comment by yogi81** [comment]
I have created a small lockscreen plugin. You can try it out yourself:
https://github.com/yogi81/screenlock_koreader_plugin

**Comment by hius07** [comment]
What happens when a user taps outside of the dialog/keyboard two times?

**Comment by yogi81** [comment]
> What happens when a user taps outside of the dialog/keyboard two times?

On kobo sage, nothing happens. The reader is blocked with a modal dialog, till the user provides a correct password. It is not possible to avoid it with simple clicking outside.

**Comment by Commodore64user** [comment]
if you switch to using an `InputContainer` instead of a `WidgetContainer` NT users could [potentially] use this as well.

**Comment by ironhak** [comment]
> I have created a small lockscreen plugin. You can try it out yourself: https://github.com/yogi81/screenlock_koreader_plugin

Amazing!!! Will you do a PR and this will hopefully inserted on Koreder?

**Comment by shuvashish76** [comment]
@pazos Please add the `user plugin` label.

**Comment by m1khal3v** [comment]
I have created one more plugin. You can try it here 
https://github.com/m1khal3v/koreader-pinlock

**Comment by oleasteo** [comment]
Since the two already mentioned plugins didn't quite satisfy me, and this seems to be the somewhat right place to share…

https://github.com/oleasteo/koreader-screenlockpin

In particular, it does support variable length PINs, lock on boot, lock on wakeup, on-the-fly unlock (no enter button), and brute-force throttling. It draws on top of the wallpaper with minimal refresh regions. Have fun! 🎉 I'm always open to feedback and ideas (over there, please don't spam here).
